packages/react-aria-components/src/DropZone.tsx

```diff
@@ -0,0 +1,160 @@
+import {DropOptions, mergeProps, useButton, useClipboard, useDrop, useFocusRing, useHover, useId, usePress, VisuallyHidden} from 'react-aria';
+import {InputContext} from './Input';
+import {isVirtualDragging} from '@react-aria/dnd';
+import {LinkContext} from './Link';
+import React, {createContext, ForwardedRef, forwardRef, useMemo, useRef} from 'react';
+import {TextContext} from './Text';
+import {useHasChild} from '@react-spectrum/utils';
+import {useLabels} from '@react-aria/utils';
+
+export interface DropZoneRenderProps {
+  /**
+   * Whether the dropzone is currently hovered with a mouse.
+   * @selector [data-hovered]
+   */
+  isHovered: boolean,
+  /**
+   * Whether the dropzone is focused, either via a mouse or keyboard.
+   * @selector :focus
+   */
+  isFocused: boolean,
+  /**
+   * Whether the dropzone is keyboard focused.
+   * @selector [data-focus-visible]
+   */
+  isFocusVisible: boolean,
+  /**
+   * Whether the dropzone is the drop target.
+   * @selector [data-drop-target]
+   */
+  isDropTarget: boolean
+}
+// note: possibly add isDisabled prop in the future
+export interface DropZoneProps extends Omit<DropOptions, 'getDropOperationForPoint' | 'onInsert' | 'onRootDrop' | 'onItemDrop' | 'onReorder'>, RenderProps<DropZoneRenderProps>, SlotProps {}
+
+export const DropZoneContext = createContext<ContextValue<DropZoneProps, HTMLDivElement>>(null);
+
+function DropZone(props: DropZoneProps, ref: ForwardedRef<HTMLDivElement>) {
+  [props, ref] = useContextProps(props, ref, DropZoneContext);
+  let {dropProps, isDropTarget} = useDrop({...props, ref});
+  let {hoverProps, isHovered} = useHover({});
+  let {focusProps, isFocused, isFocusVisible} = useFocusRing({within: true});
+
+  let buttonRef = useRef<HTMLButtonElement>(null);
+  let inputRef = useRef<HTMLInputElement>(null);
+  let uploadLinkRef = useRef<HTMLAnchorElement>(null);
+  let uploadButtonRef = useRef<HTMLButtonElement>(null);
+  let hasButton = useHasChild('button[slot=file]', ref);
+  let hasLink = useHasChild('span[slot=file]', ref);
+
+  let hasInput = useHasChild('input[type=file]', ref);
+  let textId = useId();
+  let isVirtualDrag = isVirtualDragging();
+  let labelText = useMemo(() => {
+    if (!isVirtualDrag && hasInput) {
+      return 'Press enter to select a file';
+    }
+  }, [isVirtualDrag, hasInput]);
+  let labelProps = useLabels({'aria-label': labelText, 'aria-labelledby': textId});
+
+  let {buttonProps} = useButton({
+    onPress: () => inputRef.current?.click()},
+    buttonRef);
+
+  let {clipboardProps} = useClipboard({
+    onPaste: (items) => props.onDrop?.({  
+      type: 'drop',
+      items,
+      x: 0,
+      y: 0,
+      dropOperation: 'copy'
+    })
+  });
+
+  let renderProps = useRenderProps({ 
+    ...props,
+    values: {isHovered, isFocused, isFocusVisible, isDropTarget},
+    defaultClassName: 'react-aria-DropZone'
+  });
```

Comment:
With the update to useRenderProps from https://github.com/adobe/react-spectrum/pull/4520, you'll need to add `filterDOMProps` to propagate `data-*` attributes to the DropZone div. Maybe handled by usePress already?